src/chkreg/chkreg.c

```diff
@@ -0,0 +1,447 @@
+}
+
+// Subroutine module_start - Address 0x00000248
+s32 _sceChkregInit(SceSize args, const void *argp)
+{
+    (void)args;
+    (void)argp;
+
+    // 0x00000258 - 0x0000026C
+    u32 i;
+    for (i = 0; i < CHKREG_ID_STORAGE_UMD_CONFIG_SIZE; i++)
+    {
+        g_pIdStorageUMDConfig[i] = 0;
+    }
+
+    // 0x00000270 - 0x0000028C
+    for (i = 0; i < sizeof g_ConsoleIdCertificate; i++)
+    {
+        ((u8 *)&g_ConsoleIdCertificate)[i] = 0;
+    }
+
+    g_UMDRegionCodeInfoPostIndex = 0; // 0x0x000002B0
+    g_isUMDRegionCodesObtained = SCE_FALSE;
+    g_isConsoleIdCertificateObtained = SCE_FALSE;
+
+    /* Create a module semaphore which acts as a mutex. */
+    g_semaId = sceKernelCreateSema("SceChkreg", SCE_KERNEL_SA_THFIFO, 1, 1, NULL); // 0x000002BC
+
+    /* If the semaphore could not be created successfully, unload the module. Otherwise keep it in memory. */
+    return (g_semaId < 1)
+        ? SCE_KERNEL_NO_RESIDENT
+        : SCE_KERNEL_RESIDENT;
+}
+
+// Subroutine module_stop - Address 0x000002E0
+s32 _sceChkregEnd(SceSize args, const void *argp)
+{
+    (void)args;
+    (void)argp;
+
+    s32 status;
+    SceUInt timeout;
+
+    timeout = 1 * 1000 * 1000;
+
+    // 0x00000300 - 0x00000314
+    u32 i;
+    for (i = 0; i < CHKREG_ID_STORAGE_UMD_CONFIG_SIZE; i++)
+    {
+        g_pIdStorageUMDConfig[i] = 0;
+    }
+
+    // 0x00000318 - 0x00000334
+    for (i = 0; i < sizeof g_ConsoleIdCertificate; i++)
+    {
+        ((u8 *)&g_ConsoleIdCertificate)[i] = 0;
+    }
+
+    g_UMDRegionCodeInfoPostIndex = 0; // 0x00000348
+    g_isUMDRegionCodesObtained = SCE_FALSE;
+    g_isConsoleIdCertificateObtained = SCE_FALSE;
+
+    /* Only stop the module when it is not currently used by other modules. */
+
+    status = sceKernelWaitSema(g_semaId, 1, &timeout); // 0x0x0000035C
+    if (status == SCE_ERROR_OK) // 0x0x00000364
+    {
+        /* Module no longer in use. Stop it. */
+        sceKernelDeleteSema(g_semaId); // 0x00000380
+
+        return SCE_KERNEL_STOP_SUCCESS;
+    }
+
+    /* Module still in use, don't stop it yet. */
+
+    return SCE_KERNEL_STOP_FAIL;
+}
+
+// Subroutine sceChkreg_driver_54495B19 - Address 0x00000390
+s32 sceChkregCheckRegion(u32 umdMediaType, u32 regionId)
+{
+    s32 status1;
+    s32 status2;
+
+    status1 = SCE_ERROR_SEMAPHORE; // 0x000003BC & 0x000003C8
+
+    /* Allow only one access at a time. */
+    status2 = sceKernelWaitSema(g_semaId, 1, NULL); // 0x000003C4
+    if (status2 == SCE_ERROR_OK)
+    {
+        if (g_isUMDRegionCodesObtained || (status1 = _sceChkregLookupUMDRegionCodeInfo()) == SCE_ERROR_OK) // 0x000003D8 & 0x000003E8
+        {
+            status1 = _sceChkregCheckRegion(0x80000000, umdMediaType | regionId); // 0x000003F4
+        }
+
+        /* Release acquired sema resource. */
+        status2 = sceKernelSignalSema(g_semaId, 1); // 0x00000404
+        if (status2 != SCE_ERROR_OK)
+        {
+            status1 = SCE_ERROR_SEMAPHORE;
+        }
+    }
+
+    return status1;
```

Comment:
Double status1/status2 seems a bit circumvoluted.
I'd do something like

```
s32 status;
if (sceKernelWaitSema() != SCE_ERROR_OK) {
    return SCE_ERROR_SEMAPHORE;
}
if (g_isUMD... || (status = ...) == SCE_ERROR_OK) {
    status = _sceChk...();
}
if (sceKernelSignalSema(...) != SCE_ERROR_OK) {
    return SCE_ERROR_SEMAPHORE;
}
return status;
```

Reply:
I'm not really a fan of mixing a call and a (return value) check together in one statement, but since I've wrote code like that in some places in the module already, I guess I can go ahead and make this change.